### PR TITLE
[WIP] Cache added in AWSMachine and AWSCluster reconcilers 

### PR DESCRIPTION
**Pull request by shivi28**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adding a cache in AWSCluster and AWSMachine reconcilers to avoid multiple reconcile request within a debouncing window.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2794 

**Special notes for your reviewer**:
If multiple requests comes within debouncing window then will put those requests in a queue and serve one request in each debouncing window. 

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Added cache in AWSMachine and AWSCluster reconcilers to avoid processing of multiple request within same debouncing window.
```

### CODE REVIEW

**Comment by richardcase** [comment]
@shivi28 - what happens when you requeue during reconciliation (which may be within the debounce period)?

**Comment by shivi28** [comment]
> @shivi28 - what happens when you requeue during reconciliation (which may be within the debounce period)?

Hey @richardcase, if you get multiple requests within debouncing window, then will park other requests and serve it after current debounce period overs

File: controllers/awsmachine_controller_test.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ import (
 	"github.com/golang/mock/gomock"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	. "github.com/onsi/gomega/gstruct"
+	"github.com/onsi/gomega/gstruct"
```
[the remark this answers is not in the document]
Comment (shivi28) [reply controllers/awsmachine_controller_test.go:28]: I have defined one `Options` struct in current project. But if I use dot import with `github.com/onsi/gomega/gstruct`  then it results in re-declaration of `Options` struct as it's present in gstruct lib also. That's why I switched to direct import

File: pkg/coalescing/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (shivi28) [reply pkg/coalescing/reconciler.go:2]: sure 👍

File: controllers/awsmachine_controller.go [hunk 3]
```diff
@@ -203,12 +204,17 @@ func (r *AWSMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 }
 
-func (r *AWSMachineReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+func (r *AWSMachineReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options Options) error {
 	log := ctrl.LoggerFrom(ctx)
 	AWSClusterToAWSMachines := r.AWSClusterToAWSMachines(log)
 
+	var acr reconcile.Reconciler = r
```
[the remark this answers is not in the document]
Comment (shivi28) [reply controllers/awsmachine_controller.go:211]: IMO `var acr reconcile.Reconciler = r` justify the significance of `acr` and reduces the confusion 🤔

[the remark this answers is not in the document]
Comment (Ankitasw) [reply controllers/awsmachine_controller.go:211]: I think it's already implied by r which is of type reconcile.Reconcile

**Comment by Ankitasw** [comment]
@shivi28 i think you missed release note as well, since this is a feature, it will be good to put note

**Comment by sedefsavas** [comment]
In CAPZ, they have something called coalescing reconciler which is LRU cache with a ttl so if there was a successful reconcile, reconciler result is added to the cache with a ttl. So, any reconcile requests within that ttl will not go through actual reconciliation.

Steps: (thanks @shysank)
1- reconciler req
2- check if cache (ttl cache) has a value for this request
3- if yes, return
4- if no, do normal reconciliation
5- if success -> add/update cache
6- if error, just return normally

I haven't reviewed the PR yet, but this is for reference. Would be good to have comparison between two approaches.

**Comment by enxebre** [comment]
Hey I'm a bit confused here, if I'm reading correctly the original [issue description](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2794#issue-1006464146) describes a problem that happens because all Machines are requeued all of a sudden in bulk, creating multiple different Requests and so overloading AWS API (As I read it is an scalability issue that'd you face as well if e.g you try to scale out very aggressively in one go, e.g going from 1 to 500).
How does introducing a rate limiter for individual Requests helps to solve that problem?

**Comment by CecileRobertMichon** [comment]
In CAPZ we added the coalescing reconciler in an effort to not reconcile the same object too often when reconciliation is successful. If the goal is to limit reconciles for many different Machines being queued at the same time I don't think it accomplishes that goal.

cc @devigned who was the original author of the cache + coalescing reconciler in CAPZ
ref: https://github.com/devigned/cluster-api-provider-azure/commit/b6b38b0e6f81b57e98f9eb933e3e6a9cbf2f5936

Side note: when a PR is heavily inspired from existing code in another repo please consider giving credit to the original author @shivi28

**Comment by devigned** [comment]
@CecileRobertMichon summed it up well. 

We introduced async reconciliation of resources (not waiting on the API to complete a long-running operation, like creating a VM), but rather checking back after a bit in a future reconciliation loop to see if the resource was done provisioning. With async reconciliation in some cases reconciliations would take 200ms, rather than 60+ seconds for a given resource. Since reconciliation loops were taking ~200ms they could responds to events much faster, and issue significantly more API calls / second to Azure. For example, if a controller listened to events from other resources, the activity from a large number of events could cause a high number of calls into the Azure API. The increase in Azure API calls / second caused throttling for us. That's why we introduced a minimum amount of time before a resource is allowed to reconcile after a successful reconciliation, a debouncer. Our key for the TTLRU cache was specific to a single resource.

**Comment by sedefsavas** [comment]
@CecileRobertMichon @devigned thanks for the feedbacks.

> We introduced async reconciliation of resources (not waiting on the API to complete a long-running operation, like creating a VM), but rather checking back after a bit in a future reconciliation loop to see if the resource was done provisioning. 

@devigned We considered doing that but this is not possible for CAPA as there is no Futures support in AWS API as in Azure.

**Comment by shivi28** [comment]
Thanks everyone for your reviews and apologies for delay in response due to async discussions🙏 

This PR puts the limit on reconciliation of same object, as pointed out by @CecileRobertMichon this won't help us in limiting the reconciles for many different Machines being queued at the same time.

@enxebre sorry for the confusion, but this PR doesn't solve the original issue and I raised it as an initial step to solve the rate limiting problem by introducing debouncing window in CAPA.
But after having a PR walkthrough with @sedefsavas and @pydctw, we have arrived on a conclusion to create a separate discussion on solving the original issue e2e and till then keeping this PR on hold.

**Comment by richardcase** [comment]
> @CecileRobertMichon @devigned thanks for the feedbacks.
> 
> > We introduced async reconciliation of resources (not waiting on the API to complete a long-running operation, like creating a VM), but rather checking back after a bit in a future reconciliation loop to see if the resource was done provisioning.
> 
> @devigned We considered doing that but this is not possible for CAPA as there is no Futures support in AWS API as in Azure.


In a few recent changes we have requeued instead of calling a `Wait*` style API.

**Pull request closed without merging**